File: mcp-server/test/locationSearch.test.ts
import { describe, expect, it, vi } from "vitest";
import { searchAllLocations } from "../src/locationSearch.js";
import { OpenMeteoError } from "../src/openMeteoClient.js";
import { NominatimError } from "../src/nominatimClient.js";
import type { Location } from "../src/types.js";

const { searchLocationsMock } = vi.hoisted(() => ({ searchLocationsMock: vi.fn() }));
const { searchAddressesMock } = vi.hoisted(() => ({ searchAddressesMock: vi.fn() }));

vi.mock("../src/openMeteoClient.js", async (importOriginal) => {
  const actual = await importOriginal<typeof import("../src/openMeteoClient.js")>();
  return { ...actual, searchLocations: searchLocationsMock };
});

vi.mock("../src/nominatimClient.js", async (importOriginal) => {
  const actual = await importOriginal<typeof import("../src/nominatimClient.js")>();
  return { ...actual, searchAddresses: searchAddressesMock };
});

const austinFromOpenMeteo: Location = {
  name: "Austin",
  latitude: 30.2711,
  longitude: -97.7437,
  admin1: "Texas",
  country: "US",
  manual: false,
};

const addressFromNominatim: Location = {
  name: "500 East Cesar Chavez Street",
  latitude: 30.262437,
  longitude: -97.740912,
  admin1: "Austin, Texas",
  country: "United States",
  manual: false,
};

describe("searchAllLocations", () => {
  it("merges results from both sources", async () => {
    searchLocationsMock.mockResolvedValue([austinFromOpenMeteo]);
    searchAddressesMock.mockResolvedValue([addressFromNominatim]);

    const results = await searchAllLocations("Austin");

    expect(results).toHaveLength(2);
    expect(results.map((l) => l.name)).toEqual(["Austin", "500 East Cesar Chavez Street"]);
  });

  it("collapses two results that resolve to essentially the same point", async () => {
    searchLocationsMock.mockResolvedValue([austinFromOpenMeteo]);
    searchAddressesMock.mockResolvedValue([
      { ...austinFromOpenMeteo, latitude: 30.2712, longitude: -97.7436, country: "United States" },
    ]);

    const results = await searchAllLocations("Austin");

    expect(results).toHaveLength(1);
    // Open-Meteo's copy is kept (placed first).
    expect(results[0]?.country).toBe("US");
  });

  it("returns just the successful source's results when the other rejects", async () => {
    searchLocationsMock.mockResolvedValue([austinFromOpenMeteo]);
    searchAddressesMock.mockRejectedValue(new NominatimError("Address search failed (HTTP 503)."));

    const results = await searchAllLocations("Austin");

    expect(results).toEqual([austinFromOpenMeteo]);
  });

  it("throws the Open-Meteo error when both sources reject", async () => {
    searchLocationsMock.mockRejectedValue(new OpenMeteoError("Location search failed (HTTP 503)."));
    searchAddressesMock.mockRejectedValue(new NominatimError("Address search failed (HTTP 503)."));

    await expect(searchAllLocations("Austin")).rejects.toThrow(OpenMeteoError);
  });
});

File: test/location_search_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:historical_weather/services/location_search.dart';
import 'package:historical_weather/services/nominatim_service.dart';
import 'package:historical_weather/services/open_meteo_service.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';

void main() {
  test('merges results from both sources', () async {
    final openMeteo = OpenMeteoService(
      client: MockClient((_) async => http.Response(
            jsonEncode({
              'results': [
                {'name': 'Austin', 'latitude': 30.27, 'longitude': -97.74, 'admin1': 'Texas', 'country': 'US'},
              ],
            }),
            200,
          )),
    );
    final nominatim = NominatimService(
      client: MockClient((_) async => http.Response(
            jsonEncode([
              {
                'lat': '30.2624370',
                'lon': '-97.7409120',
                'name': '',
                'address': {'house_number': '500', 'road': 'East Cesar Chavez Street', 'city': 'Austin', 'state': 'Texas', 'country': 'US'},
              },
            ]),
            200,
          )),
    );

    final results = await searchAllLocations(openMeteo: openMeteo, nominatim: nominatim, query: 'Austin');

    expect(results, hasLength(2));
    expect(results.map((l) => l.name), containsAll(['Austin', '500 East Cesar Chavez Street']));
  });

  test('collapses two results that resolve to essentially the same point', () async {
    final openMeteo = OpenMeteoService(
      client: MockClient((_) async => http.Response(
            jsonEncode({
              'results': [
                {'name': 'Austin', 'latitude': 30.2711, 'longitude': -97.7437, 'admin1': 'Texas', 'country': 'US'},
              ],
            }),
            200,
          )),
    );
    final nominatim = NominatimService(
      client: MockClient((_) async => http.Response(
            jsonEncode([
              {
                'lat': '30.2712',
                'lon': '-97.7436',
                'name': 'Austin',
                'address': {'city': 'Austin', 'state': 'Texas', 'country': 'United States'},
              },
            ]),
            200,
          )),
    );

    final results = await searchAllLocations(openMeteo: openMeteo, nominatim: nominatim, query: 'Austin');

    expect(results, hasLength(1));
    // Open-Meteo's copy is kept (added first).
    expect(results.single.country, 'US');
  });

  test('returns just the successful source\'s results when the other throws', () async {
    final openMeteo = OpenMeteoService(
      client: MockClient((_) async => http.Response(
            jsonEncode({
              'results': [
                {'name': 'Austin', 'latitude': 30.27, 'longitude': -97.74, 'admin1': 'Texas', 'country': 'US'},
              ],
            }),
            200,
          )),
    );
    final nominatim = NominatimService(client: MockClient((_) async => http.Response('', 503)));

    final results = await searchAllLocations(openMeteo: openMeteo, nominatim: nominatim, query: 'Austin');

    expect(results, hasLength(1));
    expect(results.single.name, 'Austin');
  });

  test('throws when both sources fail', () async {
    final openMeteo = OpenMeteoService(client: MockClient((_) async => http.Response('', 503)));
    final nominatim = NominatimService(client: MockClient((_) async => http.Response('', 503)));

    await expectLater(
      () => searchAllLocations(openMeteo: openMeteo, nominatim: nominatim, query: 'Austin'),
      throwsA(isA<OpenMeteoException>()),
    );
  });
}
